### PR TITLE
Refactored adapter, Node-RED 3.1.x, added nodes, fixed issues

### PR DESCRIPTION
**Pull request by klein0r**
- https://github.com/ioBroker/ioBroker.node-red/issues/291
- https://github.com/ioBroker/ioBroker.node-red/issues/384
- https://github.com/ioBroker/ioBroker.node-red/issues/332
- https://github.com/ioBroker/ioBroker.node-red/issues/323
- https://github.com/ioBroker/ioBroker.node-red/issues/389
- https://github.com/ioBroker/ioBroker.node-red/issues/276

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,342 +1,439 @@
 ![Logo](admin/node-red.png)
-# ioBroker node-red Adapter
 
-![Number of Installations](http://iobroker.live/badges/node-red-installed.svg)
-![Number of Installations](http://iobroker.live/badges/node-red-stable.svg)
-[![NPM version](http://img.shields.io/npm/v/iobroker.node-red.svg)](https://www.npmjs.com/package/iobroker.node-red)
+# ioBroker.node-red
 
-![Test and Release](https://github.com/ioBroker/ioBroker.node-red/workflows/Test%20and%20Release/badge.svg)
-[![Translation status](https://weblate.iobroker.net/widgets/adapters/-/node-red/svg-badge.svg)](https://weblate.iobroker.net/engage/adapters/?utm_source=widget)
```
Comment (Apollon77) [inline README.md:9]: Was Weblate badge removed by intent?

File: main.js [hunk 2]
```diff
@@ -168,413 +30,609 @@ function getNodeRedEditorPath() {
...
-    redProcess = spawn('node', args);
+            // create userData directory
+            if (!fs.existsSync(this.userDataDir)) {
+                fs.mkdirSync(this.userDataDir);
+            }
 
-    redProcess.on('error', err =>
-        adapter.log.error(`catched exception from node-red:${JSON.stringify(err)}`));
+            this.generateHtml()
+                .then(() => {
```
Comment (Apollon77) [inline main.js:82]: nit: why not using await here :-)

File: package.json [hunk 3]
```diff
@@ -8,8 +8,8 @@
   },
   "contributors": [
     {
-      "name": "bluefox",
-      "email": "dogafox@gmail.com"
+      "name": "Matthias Kleine",
```
Comment (Apollon77) [inline package.json:11]: Please add youself as contributor ... not replacing, or? ;-)

File: package.json [hunk 4]
```diff
@@ -23,38 +23,57 @@
     "type": "git",
     "url": "https://github.com/ioBroker/ioBroker.node-red"
   },
+  "engines": {
+    "node": ">=16"
```
Comment (Apollon77) [inline package.json:27]: Formally needs to be >=16.4.0 because of the "node: "namespaced packages you use

**Review by Apollon77 — COMMENTED** [review]
Thank you, I added some comments

File: package.json (hunk 3 above)
Comment (klein0r) [reply package.json:11]: Bluefox is already author of the package.

File: package.json (hunk 4 above)
Comment (klein0r) [reply package.json]: Thanks

File: main.js (hunk 2 above)
Comment (klein0r) [reply main.js:82]: That was the original code - not everything was rewritten 😄

File: package.json (hunk 3 above)
Comment (Apollon77) [reply package.json:11]: ok, fair enough :-)

File: README.md (hunk 1 above)
Comment (klein0r) [reply README.md:9]: Ops, I will re-add the badges

**Comment by Apollon77** [comment]
@klein0r Cool many thanks for this huge update! Just tell when you are done then I do a final look over it and we can release that (minor or major WDYT?)

**Comment by klein0r** [comment]
@Apollon77 I'm done so far. I will respond / check further issues if this is merged. Please release a major version, since I've changed a lot (e.g. admin config requires to re-enter all passwords). See added message in the io-package.json.

**Comment by klein0r** [comment]
@Apollon77 *ping* before dependabot creates merge conflicts by end of month 😄

**Comment by Apollon77** [comment]
5.0.0 on its way. Thank you very much!

**Pull request merged**